### PR TITLE
feat(sqle): add spooled query service

### PR DESCRIPTION
**Pull request by austinsley**
Ports a spool-focused spin on the existing query service for longer-running queries from our Raleigh apps. Posts a query via queryService, polls for query status at exponentially-increasing intervals until results are ready, then requests them.

### CODE REVIEW

**Comment by emoralesb05** [comment]
P.D. This also brings a few things we need to do in the query services to reduce the amount of work that clients need to do, but most of the points are still valid on either case.

**Comment by owilliams320** [comment]
so potentially to address some of the concerns where the spooled query is used within editor perhaps this stream could be modified to return an object on every emission that contains the status, sessionId, and eventually the result set on a timed interval....This could allow for the editor to store the sessionId on the first emission and this new spooled query method could skip creating a new query session if passed an existing sessionId?

Example of the observable stream emission [click here](https://rxviz.com/v/5o1ez1Wo)

I think for the first thing mentioned @emoralesb05 i could see multi-transactions an addition to the query service itself potentially as another enhancement? with this observable editor would be able to leverage stream alongside with how multi transactions are used i think

File: src/lib/sqle/spooled-query.service.ts [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+      take(1),
+      switchMap(([id]: [string, SpooledQueryState]) =>
+        this.queryService.getQueryResult(this.connectionService.current, id).pipe(
+          map((val: IQueryResultSet) => [id, val]),
+          catchError((res: HttpErrorResponse) => {
+            if (
+              res.error.message.includes(
+                'The supplied statement could not be completed due to an internal error at a database worker node.  Please contact support at Teradata.',
+              ) &&
+              currentRuns <= MAX_RETRIES
```
Comment (emoralesb05) [inline src/lib/sqle/spooled-query.service.ts]: Retrying could be very risky when we dont know the operation. I would recommend not doing this by default.

File: src/lib/sqle/spooled-query.service.ts [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+
+  getRunningInfo(): Observable<SpooledQueryState> {
+    return this.queryStatus$;
+  }
+
+  querySystem(payload: IQueryPayload): Observable<IQueryResultSet> {
+    let currentRuns: number = 1;
+
+    return this.queryService.querySystem(this.connectionService.current, { ...payload, spooledResultSet: true }).pipe(
+      tap((res: any) => this.queryStack.push(res.id)),
```
Comment (emoralesb05) [inline src/lib/sqle/spooled-query.service.ts]: We seem to only pop the queryStack when we cancel a query, not when it finishes.

File: src/lib/sqle/spooled-query.service.ts [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+  SUBMITTED = 'SUBMITTED',
+  RESPONDING = 'RESPONDING',
+  SPOOLING = 'SPOOLING',
+  RESULT_SET_READY = 'RESULT_SET_READY',
+}
+
+@Injectable()
+export class VantageSpooledQueryService {
+  queryStatus: BehaviorSubject<SpooledQueryState> = new BehaviorSubject<SpooledQueryState>(undefined);
+  queryStatus$: Observable<SpooledQueryState> = this.queryStatus.asObservable();
```
Comment (emoralesb05) [inline src/lib/sqle/spooled-query.service.ts]: We might need to do a map for this to have an observable per execution. Thoughts?

I see potential issues when trying to run 2 queries at the same time.

File: src/lib/sqle/spooled-query.service.ts (hunk 1 above)
Comment (austinsley) [reply src/lib/sqle/spooled-query.service.ts]: Fixed

File: src/lib/sqle/spooled-query.service.ts (hunk 2 above)
Comment (austinsley) [reply src/lib/sqle/spooled-query.service.ts]: Fixed

File: src/lib/sqle/spooled-query.service.ts (hunk 3 above)
Comment (austinsley) [reply src/lib/sqle/spooled-query.service.ts]: That makes sense. We can fix it by changing `querySystem` to follow the pattern in Oseer's example up above, but that's probably for another PR.

**Comment by emoralesb05** [comment]
LGTM, we can continue to improve/polish as we use it and see more use cases.

**Pull request merged**